sanity-io/structured-content-2022
# Statically Generated Sites & webhook-based on-demand page revalidation

# Statically Generated Sites & webhook-based on-demand page revalidation

## Intent

- Update Next.js to [v12.1](https://nextjs.org/blog/next-12-1)
- [Pre-render pages at build-time](https://nextjs.org/docs/basic-features/data-fetching/get-static-props) rather than on-demand
- [On-demand revalidate pages](https://nextjs.org/blog/next-12-1#on-demand-incremental-static-regeneration-beta) using [webhooks](https://www.sanity.io/docs/webhooks)

## Description

Instead of constructing Pages on-the-fly for every user, we fetch all Routes and their associated pages, and generate them at build-time.

I've set up two webhooks [in the Sanity project's settings](https://www.sanity.io/organizations/oSyH1iET5/project/33zsuc7i/api). The enabled webhook pings the _/api/revalidate_ endpoint of [this branch's latest Preview Deploy](https://structured-content-2022-web-git-next-ssg.sanity.build/api/revalidate) whenever anything in Studio is changed. When the endpoint is called, we have to find the Routes associated with the updated content and get the `slug.current` of these Routes. Then the magic happens; we call `unstable_revalidate` on each slug, which rebuilds the relevant Routes.

This setup seems _perfect_ for Sanity. The addition of `unstable_revalidate` in Next.js v12.1 could not have been better timed. I'm quite happy with how smoothly this setup seems to work, the performance benefits are massive when compared to every user having to wait for every query on every page load and navigation.

(The disabled webhook is to be enabled when this PR is merged.)

Article pages will be handled in a separate follow-up PR once the approach taking in this PR is approved.

## Testing this PR

When testing this PR, make sure that you're on the Preview Deploy that's got the branch name, and not the latest shorthash; these are not the same, and I wasted some time being puzzled at "suddenly-broken webhooks" before I realized this. :sweat_smile: 

([/sections-testing](https://stagingadmin.structuredcontent.live/desk/route;18799de8-5ecb-4432-8c6f-8586a0f82a11) and [/shared-sections](https://stagingadmin.structuredcontent.live/desk/route;36c8bda6-e3a3-4671-a260-cdaf3dcd2e43) are useful Routes for testing)

1. Open the [Realtime Logs](https://vercel.com/sanity-io/structured-content-2022-web/EZS62cMH4C8MwZbfv3crRAsEFiNv/functions?name=api%2Frevalidate) for the latest Preview Deploy for this PR, and select "api/revalidate" in the Function dropdown
2. Open [Studio](https://stagingadmin.structuredcontent.live/) and change the following elements:
    - A sharedSection referenced by a Page
    - The contents of a Page
    - The slug and metadata of a Route
3. For each of the changes above, verify that:
    - a `[POST] /api/revalidate` request is logged in the _Realtime Logs_
    - that you get output indicating success is logged the _Realtime Logs_, e.g. `[revalidate] Updated routes: shared-sections`
    - That the Route referencing the updated element renders the updated content on refresh (this is where using [the correct Preview Deploy URL](https://structured-content-2022-web-git-next-ssg.sanity.build) is important!)
